### DISCUSSION TITLE
Max image resolution check before generation

### DISCUSSION THREAD

**Discussion by fiveseven808** [Kilvoctu/aiyabot#172] [Ideas]
I wanted to call this a bug, but it's not implemented in the A1111 webui yet... what I've discovered is for most generations (at least the ones I do), they're limited by total pixel count. i.e. I can generate a 2mp image of arbitrary aspect ratio, but pretty much everything above that will fail and I'll run out of VRAM. 

Since (to me) it is easier to implement this check on a discord bot than to figure out where to do it in the A1111 repo, AND the goal of the discord bot is to simplify things, I'd like to request it here. 

**How it would work:**
When a user implements a custom resolution, the height and width would be multiplied against each other and checked against a max pixel count (set either by channel or globally). 
- It will error out if the max pixel count exceeds the set limit OR
- It will attempt to keep the same aspect ratio by reducing the resolution until it is under the limit

**Why I want this?**
It will simplify learning how to use the bot for new users and help them understand the limits of the system it's running on 
It will reduce the memory leak effect A1111 has over time necessitating a restart of the webui to clear VRAM

**Comment by Kilvoctu** [comment]
Heya, this seems overly convoluted to me, when max_size already exists so hosts don't exceed what their hardware is capable of, and channel defaults exists so end users don't necessarily need to input anything other than a prompt. 🤔 
It to me seems to complicate drawing, asking users to do math for width and height to stay under 2MP (this reasoning is also why AIYA only supports legacy high-res fix).
 
Anyway, I'm not seeing benefit of this, aside from desire to minmax the output from your hardware, at the cost of putting the burden on the users to figure out the limits.
If your Web UI has a memory leak, I'm not sure what I can do about that... My AIYA is a separate frontend application; I don't develop the Web UI or know how it works outside of its API.

As a sidenote, the goal of the aiyabot is to simplify things specifically for the end user, not in general. I'm very okay with shifting all the work to the bot host to benefit the user. Again, I'm not sold on how this creates an improved end user experience..

**Comment by fiveseven808** [comment]
When Height and Width are an option, people will play with it. My computer can't do 2048x2048 images, but can do 2048x1024 images, which is why I wanted the option to go up to 2048. Aspect ratio has an extreme impact on the type of image generated, so being able to go up that high when doing an img2img upscale is necessary for any sort of quality. 

Since your bot returns an invalid parameter error when the webui runs out of vram, maybe it should be considered a bug since an incorrect error is being thrown back at the user? To the uneducated without access to the webui console, the discord user would have no idea what prompted the failed generation. I understand that fixing the webui is not within the scope of this project, but repeated failed high resolution/out of memory generations cause a memory leak that eventually leads to the webui running out of memory on any generation.

If a non-malicious but curious user pushes the resolution too high, and renders the image server unusable and this is a serious problem. To lower the max resolution to a level where it would never happen would be to gimp those who know how to use the tool properly.

**Reply by Kilvoctu** [reply]
This follow-up doesn't change my opinion 🤔. To me it still looks like you're exceeding what your machine is capable of because you want more than it can handle, causing the backend, and by extension AIYA, to fail. I have only a GTX1080, and I don't place blame on Web UI or my AIYA when doing a too-high-res request that breaks my applications. 

That said, I added this feature request to my [lists of projects](https://github.com/users/Kilvoctu/projects/12/views/1?pane=issue&itemId=26360429) to get around to at some point. Since my code is a bit spaghetti, it could take a while for this particular request. Also, I'm not sure what priority you want your requests in 🤔 I think you've made a few and I don't remember which ones I've been able to address.
